style.md

```diff
@@ -1515,6 +1516,114 @@ Note that the compiler will not generate errors when using predeclared
 identifiers, but tools such as `go vet` should correctly point out these and
 other cases of shadowing.
 
+### Avoid `init()`
+
+Avoid `init()` where possible. At a minimum, all code within `init()` should:
+
+1. Be completely deterministic, regardless of program environment or invocation.
+2. Be completely self-contained within the program.
+3. Not depend on side-effects of other `init()` functions.
+4. Be able to run in any order. While `init()` ordering is well-known, code can
+   change, and thus `init()` functions that rely on specific sequencing make
+   code brittle and error-prone.
+5. Not access or manipulate global or environment state, such as machine
+   information, environment variables, working directory, program
+   arguments/inputs, etc.
+6. Not perform I/O, including both filesystem, network, and system calls.
```

Comment:
I think (1), (5) and (6) have a bit of overlap, so we may be able to merge them.

I can definitely see `init` functions being useful that violate (1), (5), and (6) (e.g., creating a variable parsed from an environment var, or parsing some data). Only really used when dealing with globals typically.

Since these requirements are marked as "At a minimum, all code [..] should", I think it might be a stronger message than the typical "avoid" messaging.

Reply:
I agree, but also think that if we merge them we end up with either (a) sparse guidance that doesn't really speak to all of the elements involved or (b) a mega-bullet. I'll see if there's a better way to combine them that condenses without losing meaning.

RE "all code should" vs "avoid", agreed. We should use more permissive language.

Reply:
Reworked these a bit, let me know what you think. I left state and IO as separate points, it felt a little dense when giving examples of both, and I think they're distinct enough that separating them doesn't feel too redundant.